packages/hub/card-services.ts

```diff
@@ -0,0 +1,233 @@
+import Session from '@cardstack/plugin-utils/session';
+import { declareInjections } from '@cardstack/di';
+import { todo } from '@cardstack/plugin-utils/todo-any';
+import { SingleResourceDoc, ResourceObject, ResourceIdentifierObject, CollectionResourceDoc } from 'jsonapi-typescript';
+import { get, uniqBy } from 'lodash';
+
+const cardIdDelim = '::';
+const cardBrowserAssetFields = [
+  'isolated-template',
+  'isolated-js',
+  'isolated-css',
+  'embedded-template',
+  'embedded-js',
+  'embedded-css',
+  'edit-template',
+  'edit-js',
+  'edit-css',
+];
+interface CardContext {
+  repository: string;
+  packageName: string;
+  cardId?: string;
+  modelId?: string;
+}
+
+export = declareInjections({
+  writers: 'hub:writers',
+  indexers: 'hub:indexers',
+  searchers: 'hub:searchers',
+  currentSchema: 'hub:current-schema',
+},
+
+class CardServices {
+  writers: todo;
+  indexers: todo;
+  searchers: todo;
+  currentSchema: todo;
+
+  cardContextFromId(id: string) {
+    let [repository, packageName, cardId, modelId] = id.split(cardIdDelim);
+
+    return {
+      repository,
+      packageName,
+      cardId,
+      modelId,
+    };
+  }
+
+  cardContextToId({ repository, packageName, cardId, modelId }: CardContext) {
+    return [repository, packageName, cardId, modelId].filter(i => i != null).join(cardIdDelim);
+  }
+
+  // TODO should we care about the session of the caller who is requesting to load the card?
+  async loadCard(card: SingleResourceDoc) {
+    if (!card.data.id) { throw new Error(`Cannot load card with missing id.`); }
+    let cardId = card.data.id;
+
+    let existingCardModel;
+    let { repository, packageName } = this.cardContextFromId(cardId);
+    let cardModelType = this.cardContextToId({ repository, packageName });
+    try {
+      // use the card's model to check for presence of a card, as checking for the presence using
+      // the card itself will result in cycle as the act of searching for a card here will
+      // trigger a card to be loaded if it does not exist in the index.
+      existingCardModel = await this.searchers.get(Session.INTERNAL_PRIVILEGED, 'local-hub', cardModelType, cardId);
+    } catch (e) { if (e.status !== 404) { throw e; } }
+
+    if (existingCardModel) {
+      // we've already loaded this card so just stop. maybe in
+      // the future we can pay attention to the card version and
+      // reload if version has changed...
+      return;
+    }
+
+    let schemaModels = await this.buildCardModelSchema(card);
+    if (!schemaModels) { return; }
+
+    for (let model of schemaModels) {
+      await this.writers.create(Session.INTERNAL_PRIVILEGED, model.type, { data: model });
+    }
+    await this.writers.create(Session.INTERNAL_PRIVILEGED, 'cards', card);
```

Comment:
`writer.create()` here is a problem.

Both the card and its internal model should already be real models in some data source. (With namespaced ids and fields, presumably, so they fit into the existing system.) It doesn't make sense to create them here.

And even if these created models are needed as legitimate derived state, it would still be incorrect to write derived state into the real data sources like this.

Reply:
ok I see, i think i went down this path because I wrote the tests around getting a card before I wrote the tests around the indexers and searchers being able to auto new cards they encounter. I'll re-work this.

Reply:
@ef4 I'm working thru how to load schema discovered by a searcher without writing it to the default data source. specifically how to update the schema that is available via `hub:current-schema`. This ultimately points to hub:indexers.schema(), which is where the schema cache lives. I was thinking that if the searcher encountered a new card, it could use the schema loader to load the new card's schema models, and then call a new function on hub:indexers to update the schema cache with the newly derived schema. does this sound like a reasonable approach?

this would mean that the card schema is not necessarily persisted in the index--but I think that is probably ok, since the card schema would automatically be loaded the next time the hub starts via the same means. (like whenever the card is encountered)

Reply:
I guess then the concern would be if an indexer emitted some new schema via conventional means, would the card schema we inserted into the schema cache be lost by a new schema provided by a `RunningIndexer`? I think you mentioned this way back when we were talking thru this in july, around the idea that the schema would just have to build up over time--i think I need to accommodate preserving discovered card schema with the RunningIndexer provided schema.

Reply:
Sorry for the rubber ducking--it looks like Schema has an `applyChanges()` that lets me add additional schema models. Maybe I can keep track of the discovered card schema during teh lifetime of the hub, and just use applyChanges to inject any discovered card schema when the running indexers return more schema.